### PR TITLE
Sample Validation Data By Study

### PR DESCRIPTION
**Pull request by ben-heil**
The primary changes in this PR add three functions to `classifier.py`:
1. `get_validation_set` selects studies at random to be used as validation data until a threshold is reached
2. `prepare_input_data` moves some of the complexity of PLIER dimensionality reduction and validation data selection out of `load_data` to increase readability
3. `train_model` moves model training out of the main function in anticipation of creating a multi-model evaluation script in the future

### CODE REVIEW

File: brdnet/classifier.py [hunk 1]
```diff
@@ -111,33 +214,47 @@ def load_data(args):
...
+    # Create log directory
+    os.mkdir(args.logdir)
+    model_name = args.model
+    validate_model_name(model_name)
+
+    model = get_model(model_name, args.logdir, args.learning_rate)
 
-    return train_X, train_Y
+    model.fit(train_X, train_Y,
+            batch_size=16,
```
Comment (jjc2718) [inline brdnet/classifier.py]: Might be good to have batch size as an adjustable hyperparameter too (I've noticed in past projects of mine that it can affect performance, but YMMV)

File: brdnet/classifier.py [hunk 2]
```diff
@@ -96,6 +102,103 @@ def reduce_dimensionality(expression_df, Z_df):
...
+       Returns
+       -------
+       train_df: pandas.DataFrame
+           A dataframe containing the fraction of the sample to be used in training
+       val_df:
+           A dataframe containing the fraction of the sample to be used for validation
+    '''
+    studies = [name.split('.')[0] for name in data.columns]
+    # random.shuffle shuffles in place
+    random.shuffle(studies)
```
Comment (jjc2718) [inline brdnet/classifier.py]: Will this give a reproducible ordering of the data for a given seed? You might have to seed the `random` module in addition to the others (i.e. by calling `random.seed(args.seed)` where you call `numpy.random.seed`).

(I think they have separate internal states/random number generators, but I could be wrong)

File: brdnet/classifier.py [hunk 3]
```diff
@@ -111,33 +214,47 @@ def load_data(args):
...
-    train_X = numpy.concatenate([healthy_matrix, disease_matrix])
-    train_Y = numpy.concatenate([healthy_labels, disease_labels])
 
-    #TODO keep metadata with training data somehow
-    # (https://github.com/greenelab/brd-net/pull/7#discussion_r309284156)
+def train_model(train_X, train_Y, val_X, val_Y,
+                model_name=None, logdir=None, lr=None, epochs=None,
+                random_seed=42):
+    # Create log directory
+    os.mkdir(args.logdir)
```
Comment (jjc2718) [inline brdnet/classifier.py]: Should this be `logdir` (not `args.logdir`)?

Same question for the following line (`args.model` -> `model`)

File: brdnet/classifier.py [hunk 4]
```diff
@@ -111,33 +214,47 @@ def load_data(args):
...
+    model_name = args.model
+    validate_model_name(model_name)
+
+    model = get_model(model_name, args.logdir, args.learning_rate)
```
Comment (jjc2718) [inline brdnet/classifier.py]: `args.logdir` -> `logdir` and `args.learning_rate` -> `lr`?

File: brdnet/classifier.py [hunk 5]
```diff
@@ -111,33 +214,47 @@ def load_data(args):
...
 
-    return train_X, train_Y
+    model.fit(train_X, train_Y,
+            batch_size=16,
+            epochs=args.epochs,
+            callbacks=[tf.keras.callbacks.TensorBoard(log_dir=args.logdir),
+                       tf.keras.callbacks.ReduceLROnPlateau(min_lr=1e-11),
```
Comment (jjc2718) [inline brdnet/classifier.py]: It would be good for this to be explicitly documented somewhere (for example, in the function docstring, or in the `help` string for your `learning_rate` argument).

I know learning rate decay is becoming fairly standard for these sorts of classifiers, but it would be helpful to make sure that users know this is the default.

File: brdnet/classifier.py [hunk 6]
```diff
@@ -96,6 +102,103 @@ def reduce_dimensionality(expression_df, Z_df):
...
+    val_X: numpy.array
+        The gene expression data to be held out to evaluate model training
+    val_Y: numpy.array
+        The labels for val_X
+    '''
+    healthy_train, healthy_val = get_validation_set(healthy_df)
+    healthy_train = reduce_dimensionality(healthy_train, Z_df)
+    healthy_val = reduce_dimensionality(healthy_val, Z_df)
+
+    disease_train, disease_val = get_validation_set(disease_df)
```
Comment (jjc2718) [inline brdnet/classifier.py]: Do you want to have the same partitions for both the healthy and disease datasets? So, for instance, if study 1 and 2 are the training set and study 3 is the validation set for the healthy samples, do you want to have the same 1+2 / 3 partition for the disease samples?

It seems like the way you've written this, you could get different partitions for healthy vs. disease (but correct me if I'm wrong, or if this is intentional).

**Review by jjc2718 — APPROVED** [review]
Nice work! Just a few questions/points of clarification.

File: brdnet/classifier.py (hunk 2 above)
Comment (ben-heil) [reply brdnet/classifier.py]: Good catch, I had forgotten that I used random in addition to numpy and tf's random functions

File: brdnet/classifier.py (hunk 3 above)
Comment (ben-heil) [reply brdnet/classifier.py]: Yep, it hadn't broken yet since I hadn't called the function outside of `classifier.py`

File: brdnet/classifier.py (hunk 4 above)
Comment (ben-heil) [reply brdnet/classifier.py]: Correct

File: brdnet/classifier.py (hunk 5 above)
Comment (ben-heil) [reply brdnet/classifier.py]: Good point, I added a sentence to the help string to address it.

File: brdnet/classifier.py (hunk 2 above)
Comment (ben-heil) [reply brdnet/classifier.py]: I took a closer look at it when modifying the `get_validation_set` study and realized that shuffling the studies before putting them into a dict doesn't really do anything. I removed the shuffle line to eliminate redundancy.

File: brdnet/classifier.py (hunk 6 above)
Comment (ben-heil) [reply brdnet/classifier.py]: Yep, I totally forgot to synchronize the healthy and unhealthy studies. I modified the function to fix this.
